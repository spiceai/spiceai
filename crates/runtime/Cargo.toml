[package]
description = "The spiced runtime"
edition.workspace = true
homepage.workspace = true
license.workspace = true
name = "runtime"
repository.workspace = true
rust-version.workspace = true
version.workspace = true

[dependencies]
app = { path = "../app" }
arrow-flight = { workspace = true, features = ["flight-sql-experimental"] }
arrow-ipc.workspace = true
arrow-json.workspace = true
arrow.workspace = true
arrow-csv = "53.1.0"
arrow-schema = { workspace = true, features= ["serde"] }
arrow_sql_gen = { path = "../arrow_sql_gen" }
arrow_tools = { path = "../arrow_tools" }
async-openai.workspace = true
async-stream.workspace = true
async-trait.workspace = true
aws-config = { version = "1.1.10", optional = true }
aws-sdk-secretsmanager = { version = "1.21.0", optional = true }
aws-sdk-sts = { version = "1.19.0", optional = true }
axum = { version = "0.7.4", features = ["macros"] }
axum-extra = { version = "0.9.3", features = ["typed-header"] }
base64.workspace = true
bytes = { version = "1", default-features = false }
cache = { path = "../cache" }
chrono = { version = "0.4.38" }
clap.workspace = true
clickhouse-rs = { workspace = true, optional = true }
csv = "1.3.0"
dashmap = "5.5.3"
data_components = { path = "../data_components" }
datafusion-federation = { workspace = true }
datafusion-federation-sql = { workspace = true }
datafusion-functions-json = { workspace = true }
datafusion-table-providers = { workspace = true }
datafusion.workspace = true
db_connection_pool = { path = "../db_connection_pool" }
document_parse = { path = "../document_parse" }
dotenvy.workspace = true
duckdb = { workspace = true, features = [
  "bundled",
  "r2d2",
  "vtab",
  "vtab-arrow",
], optional = true }
flight_client = { path = "../flight_client" }
fundu = { workspace = true }
futures.workspace = true
globset.workspace = true
graph-rs-sdk = { workspace = true, optional = true }
graphql-parser.workspace = true
headers-accept = "0.1.3"
http = "1.1.0"
http-body-util = "0.1.2"
hyper = "1.4.1"
hyper-util = { version = "0.1.6", features = ["service"] }
indexmap = "2"
itertools.workspace = true
keyring = { version = "2.3.2", optional = true }
lazy_static.workspace = true
llms = { path = "../llms" }
logos = "0.14.0"
mediatype = "0.19.18"
model_components = { path = "../model_components" }
tiberius ={ workspace = true, optional = true }
mysql_async = { workspace = true, optional = true }
notify = "6.1.1"
ns_lookup = { path = "../ns_lookup" }
object_store = { workspace = true, features = ["aws", "http", "azure"] }
once_cell = "1.19.0"
opentelemetry.workspace = true
opentelemetry_sdk.workspace = true
opentelemetry-proto = { version = "0.25", features = [
  "gen-tonic-messages",
  "gen-tonic",
  "metrics",
] }
opentelemetry-prometheus.workspace = true
otel-arrow = { path = "../otel-arrow" }
pin-project = "1.0"
prometheus.workspace = true
prometheus-parse = "0.2.5"
prost = { version = "0.13.1", default-features = false, features = [
  "prost-derive",
] }
rand = "0.8.5"
regex.workspace = true
reqwest.workspace = true
rustls.workspace = true
rustls-pemfile.workspace = true
rusqlite = { workspace = true, optional = true }
schemars = "0.8.19"
secrecy.workspace = true
serde = {workspace = true, features = ["derive"] }
serde_json.workspace = true
snafu.workspace = true
snowflake-api = { workspace = true, optional = true }
spicepod = { path = "../spicepod" }
ssh2 = { workspace = true, optional = true }
suppaftp = { workspace = true, optional = true }
telemetry = { path = "../telemetry" }
tokio-rusqlite = { workspace = true, optional = true }
tokio-rustls = "0.26.0"
tokio.workspace = true
tokio-util = { workspace = true, optional = true }
tonic-health.workspace = true
tonic.workspace = true
tracing.workspace = true
tracing-futures.workspace = true
tract-core = "0.21.0"
url = "2.5.0"
util = { path = "../util" }
uuid.workspace = true
x509-certificate.workspace = true
jsonwebtoken = "9.3.0"

[dev-dependencies]
anyhow = "1.0.86"
async-graphql = "7.0.5"
async-graphql-axum = "7.0.5"
bollard = "0.16.1"
spice-cloud = { path = "../spice_cloud" }
tracing-subscriber.workspace = true
insta.workspace = true
azure_storage_blobs = "0.20.0"
azure_core = "0.20.0"
azure_storage = "0.20.0"

[features]
aws-secrets-manager = [
  "dep:aws-config",
  "dep:aws-sdk-secretsmanager",
  "dep:aws-sdk-sts",
]
clickhouse = [
  "dep:clickhouse-rs",
  "db_connection_pool/clickhouse",
  "arrow_sql_gen/clickhouse",
  "data_components/clickhouse",
]
databricks = ["data_components/databricks"]
debezium = ["data_components/debezium"]
default = ["keyring-secret-store", "aws-secrets-manager", "sharepoint"]
delta_lake = ["data_components/delta_lake"]
dev = []
dremio = []
duckdb = ["dep:duckdb", "db_connection_pool/duckdb", "data_components/duckdb"]
flightsql = ["data_components/flightsql"]
ftp = ["dep:suppaftp", "dep:ssh2"]
keyring-secret-store = ["dep:keyring"]
models = ["model_components/full", "llms/mistralrs", "datafusion-table-providers/postgres"]
mssql = ["dep:tiberius", "dep:tokio-util", "data_components/mssql"]
mysql = ["dep:mysql_async", "db_connection_pool/mysql", "data_components/mysql"]
odbc = ["db_connection_pool/odbc", "data_components/odbc"]
postgres = [
  "db_connection_pool/postgres",
  "data_components/postgres",
  "datafusion-table-providers/postgres",
]
sharepoint = ["data_components/sharepoint", "dep:graph-rs-sdk"]
snowflake = [
  "dep:snowflake-api",
  "db_connection_pool/snowflake",
  "data_components/snowflake",
]
spark = ["data_components/spark_connect"]
sqlite = [
  "dep:tokio-rusqlite",
  "db_connection_pool/sqlite",
  "data_components/sqlite",
  "dep:rusqlite",
]

[[bench]]
harness = false
name = "bench"
