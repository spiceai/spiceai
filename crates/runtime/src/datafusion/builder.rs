/*
Copyright 2024 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use std::{
    collections::HashSet,
    sync::{Arc, RwLock},
};

use cache::QueryResultsCacheProvider;
use datafusion::{
    catalog_common::{CatalogProvider, MemoryCatalogProvider},
    execution::SessionStateBuilder,
    prelude::{SessionConfig, SessionContext},
};
use datafusion_federation::FederationAnalyzerRule;
use tokio::sync::RwLock as TokioRwLock;

use crate::{embeddings, object_store_registry::default_runtime_env, status};

use super::{
    extension::{bytes_processed::BytesProcessedOptimizerRule, SpiceQueryPlanner},
    schema::SpiceSchemaProvider,
    DataFusion, SPICE_DEFAULT_CATALOG, SPICE_DEFAULT_SCHEMA, SPICE_METADATA_SCHEMA,
    SPICE_RUNTIME_SCHEMA,
};

pub struct DataFusionBuilder {
    config: SessionConfig,
    status: Arc<status::RuntimeStatus>,
    cache_provider: Option<Arc<QueryResultsCacheProvider>>,
}

impl DataFusionBuilder {
    #[must_use]
    pub fn new(status: Arc<status::RuntimeStatus>) -> Self {
        let mut df_config = SessionConfig::new()
            .with_information_schema(true)
            .with_create_default_catalog_and_schema(false);

        df_config.options_mut().sql_parser.dialect = "PostgreSQL".to_string();
        df_config.options_mut().catalog.default_catalog = SPICE_DEFAULT_CATALOG.to_string();
        df_config.options_mut().catalog.default_schema = SPICE_DEFAULT_SCHEMA.to_string();
        df_config.options_mut().execution.keep_partition_by_columns = true;
        df_config
            .options_mut()
            .execution
            .listing_table_ignore_subdirectory = false;

        Self {
            config: df_config,
            status,
            cache_provider: None,
        }
    }

    #[must_use]
    pub fn with_cache_provider(mut self, cache_provider: Arc<QueryResultsCacheProvider>) -> Self {
        self.cache_provider = Some(cache_provider);
        self
    }

    #[must_use]
    pub fn keep_partition_by_columns(mut self, keep_partition_by_columns: bool) -> Self {
        self.config
            .options_mut()
            .execution
            .keep_partition_by_columns = keep_partition_by_columns;
        self
    }

    /// Builds the `DataFusion` instance.
    ///
    /// # Panics
    ///
    /// Panics if the `DataFusion` instance cannot be built due to errors in registering functions or schemas.
    #[must_use]
    pub fn build(self) -> DataFusion {
        let mut state = SessionStateBuilder::new()
            .with_config(self.config)
            .with_default_features()
            .with_query_planner(Arc::new(SpiceQueryPlanner::new()))
            .with_runtime_env(default_runtime_env())
            .build();

        if let Err(e) = datafusion_functions_json::register_all(&mut state) {
            panic!("Unable to register JSON functions: {e}");
        };

        let ctx = SessionContext::new_with_state(state);
        ctx.add_analyzer_rule(Arc::new(FederationAnalyzerRule::new()));
        ctx.add_optimizer_rule(Arc::new(BytesProcessedOptimizerRule::new()));
        ctx.register_udf(embeddings::cosine_distance::CosineDistance::new().into());
        ctx.register_udf(crate::datafusion::udf::Greatest::new().into());
        ctx.register_udf(crate::datafusion::udf::Least::new().into());
        let catalog = MemoryCatalogProvider::new();
        let default_schema = SpiceSchemaProvider::new();
        let runtime_schema = SpiceSchemaProvider::new();
        let metadata_schema = SpiceSchemaProvider::new();

        match catalog.register_schema(SPICE_DEFAULT_SCHEMA, Arc::new(default_schema)) {
            Ok(_) => {}
            Err(e) => {
                panic!("Unable to register default schema: {e}");
            }
        }

        match catalog.register_schema(SPICE_RUNTIME_SCHEMA, Arc::new(runtime_schema)) {
            Ok(_) => {}
            Err(e) => {
                panic!("Unable to register spice runtime schema: {e}");
            }
        }

        match catalog.register_schema(SPICE_METADATA_SCHEMA, Arc::new(metadata_schema)) {
            Ok(_) => {}
            Err(e) => {
                panic!("Unable to register spice runtime schema: {e}");
            }
        }

        ctx.register_catalog(SPICE_DEFAULT_CATALOG, Arc::new(catalog));

        DataFusion {
            runtime_status: self.status,
            ctx: Arc::new(ctx),
            data_writers: RwLock::new(HashSet::new()),
            cache_provider: RwLock::new(self.cache_provider),
            pending_sink_tables: TokioRwLock::new(Vec::new()),
            accelerated_tables: TokioRwLock::new(HashSet::new()),
        }
    }
}
