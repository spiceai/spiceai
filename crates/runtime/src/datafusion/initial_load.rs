/*
Copyright 2024 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use std::sync::MutexGuard;

use crate::DataFusion;

impl DataFusion {
    pub fn mark_initial_load_complete(&self) {
        let mut guard = self.get_initial_load_guard();
        *guard = true;
    }

    pub fn is_initial_load_complete(&self) -> bool {
        let guard = self.get_initial_load_guard();
        *guard
    }

    fn get_initial_load_guard(&self) -> MutexGuard<bool> {
        match self.initial_load_complete.lock() {
            Ok(guard) => guard,
            Err(poisoned) => poisoned.into_inner(),
        }
    }
}
