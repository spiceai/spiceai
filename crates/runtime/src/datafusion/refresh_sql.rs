/*
Copyright 2024 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use datafusion::sql::parser::{DFParser, Statement};
use datafusion::sql::sqlparser;
use datafusion::sql::sqlparser::ast::SetExpr;
use datafusion::sql::sqlparser::dialect::PostgreSqlDialect;
use snafu::prelude::*;
use sqlparser::ast::Statement as SQLStatement;

pub type Result<T, E = Error> = std::result::Result<T, E>;

#[derive(Debug, Snafu)]
pub enum Error {
    #[snafu(display("Unable to parse the refresh SQL: {source}"))]
    UnableToParseSql {
        source: sqlparser::parser::ParserError,
    },

    #[snafu(display(
        "Expected a single SQL statement for the refresh SQL, found {num_statements}"
    ))]
    ExpectedSingleSqlStatement { num_statements: usize },

    #[snafu(display("Expected a SQL query starting with SELECT * FROM {expected_table}"))]
    InvalidSqlStatement { expected_table: String },

    #[snafu(display("Missing expected SQL statement - this is a bug in Spice.ai"))]
    MissingStatement,
}

#[allow(clippy::module_name_repetitions)]
pub fn validate_refresh_sql(expected_table: &str, refresh_sql: &str) -> Result<()> {
    let mut statements = DFParser::parse_sql_with_dialect(refresh_sql, &PostgreSqlDialect {})
        .context(UnableToParseSqlSnafu)?;
    if statements.len() != 1 {
        ExpectedSingleSqlStatementSnafu {
            num_statements: statements.len(),
        }
        .fail()?;
    }

    let statement = statements.pop_front().context(MissingStatementSnafu)?;
    match statement {
        Statement::Statement(statement) => match statement.as_ref() {
            SQLStatement::Query(query) => match query.body.as_ref() {
                SetExpr::Select(select) => {
                    ensure!(
                        select.projection.len() == 1,
                        InvalidSqlStatementSnafu { expected_table }
                    );
                    ensure!(
                        matches!(
                            select.projection[0],
                            sqlparser::ast::SelectItem::Wildcard(_)
                        ),
                        InvalidSqlStatementSnafu { expected_table }
                    );
                    ensure!(
                        select.from.len() == 1,
                        InvalidSqlStatementSnafu { expected_table }
                    );

                    match &select.from[0].relation {
                        sqlparser::ast::TableFactor::Table {
                            name,
                            alias: _,
                            args: _,
                            with_hints: _,
                            version: _,
                            partitions: _,
                        } => {
                            ensure!(
                                name.0.len() == 1,
                                InvalidSqlStatementSnafu { expected_table }
                            );
                            ensure!(
                                name.0[0].value.as_str() == expected_table,
                                InvalidSqlStatementSnafu { expected_table }
                            );
                        }
                        _ => {
                            InvalidSqlStatementSnafu { expected_table }.fail()?;
                        }
                    }

                    Ok(())
                }
                _ => InvalidSqlStatementSnafu { expected_table }.fail()?,
            },
            _ => InvalidSqlStatementSnafu { expected_table }.fail()?,
        },
        _ => InvalidSqlStatementSnafu { expected_table }.fail()?,
    }
}
