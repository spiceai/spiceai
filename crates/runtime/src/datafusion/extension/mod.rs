/*
Copyright 2024 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use async_trait::async_trait;
use datafusion::{
    error::Result,
    execution::context::{QueryPlanner, SessionState},
    logical_expr::{LogicalPlan, UserDefinedLogicalNode},
    physical_plan::ExecutionPlan,
    physical_planner::{DefaultPhysicalPlanner, ExtensionPlanner, PhysicalPlanner},
};
use datafusion_federation::FederatedPlanner;
use std::sync::Arc;
use telemetry::{BytesScannedExec, BytesScannedNode};

pub mod analyzer;
pub mod telemetry;

#[derive(Default)]
pub struct SpiceQueryPlanner {}

impl SpiceQueryPlanner {
    pub fn new() -> Self {
        SpiceQueryPlanner {}
    }
}

#[async_trait]
impl QueryPlanner for SpiceQueryPlanner {
    async fn create_physical_plan(
        &self,
        logical_plan: &LogicalPlan,
        session_state: &SessionState,
    ) -> Result<Arc<dyn ExecutionPlan>> {
        let physical_planner = DefaultPhysicalPlanner::with_extension_planners(vec![
            Arc::new(FederatedPlanner::new()),
            Arc::new(SpiceExtensionPlanner::new()),
        ]);
        physical_planner
            .create_physical_plan(logical_plan, session_state)
            .await
    }
}

#[derive(Default)]
pub struct SpiceExtensionPlanner {}

impl SpiceExtensionPlanner {
    pub fn new() -> Self {
        SpiceExtensionPlanner {}
    }
}

#[async_trait]
impl ExtensionPlanner for SpiceExtensionPlanner {
    async fn plan_extension(
        &self,
        _planner: &dyn PhysicalPlanner,
        node: &dyn UserDefinedLogicalNode,
        logical_inputs: &[&LogicalPlan],
        physical_inputs: &[Arc<dyn ExecutionPlan>],
        _session_state: &SessionState,
    ) -> Result<Option<Arc<dyn ExecutionPlan>>> {
        // bytes_scanned Telemetry Extension
        let bytes_scanned_node = node.as_any().downcast_ref::<BytesScannedNode>();
        if bytes_scanned_node.is_some() {
            assert_eq!(logical_inputs.len(), 1, "should have 1 input");
            assert_eq!(physical_inputs.len(), 1, "should have 1 input");
            let Some(physical_input) = physical_inputs.iter().next() else {
                panic!("should have 1 input");
            };

            let exec_plan = Arc::new(BytesScannedExec::new(Arc::clone(physical_input)));
            return Ok(Some(exec_plan));
        }
        Ok(None)
    }
}
