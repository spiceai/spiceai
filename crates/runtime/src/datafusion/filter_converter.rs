use arrow::datatypes::DataType;
use datafusion::{
    logical_expr::{binary_expr, cast, col, lit, Expr, Operator},
    scalar::ScalarValue,
};
use spicepod::component::dataset::TimeFormat;

#[derive(Debug, Clone)]
enum ExprTimeFormat {
    ISO8601,
    UnixTimestamp(ExprUnixTimestamp),
    Timestamp,
}

#[derive(Debug, Clone)]
struct ExprUnixTimestamp {
    scale: u128,
}

#[derive(Clone, Debug)]
pub(crate) struct TimestampFilterConvert {
    time_column: String,
    time_format: ExprTimeFormat,
}

#[allow(clippy::needless_pass_by_value)]
impl TimestampFilterConvert {
    pub(crate) fn create(
        field: Option<arrow::datatypes::Field>,
        time_column: Option<String>,
        mut time_format: Option<TimeFormat>,
    ) -> Option<Self> {
        let field = field?;
        let time_column = time_column?;

        let time_format = match field.data_type() {
            DataType::Int8
            | DataType::Int16
            | DataType::Int32
            | DataType::Int64
            | DataType::UInt8
            | DataType::UInt16
            | DataType::UInt32
            | DataType::UInt64
            | DataType::Float16
            | DataType::Float32
            | DataType::Float64 => {
                let mut scale = 1_000_000_000;
                if let Some(time_format) = time_format.take() {
                    if time_format == TimeFormat::UnixMillis {
                        scale = 1_000_000;
                    }
                }
                ExprTimeFormat::UnixTimestamp(ExprUnixTimestamp { scale })
            }
            DataType::Timestamp(_, _)
            | DataType::Date32
            | DataType::Date64
            | DataType::Time32(_)
            | DataType::Time64(_) => ExprTimeFormat::Timestamp,
            DataType::Utf8 | DataType::LargeUtf8 => ExprTimeFormat::ISO8601,
            _ => {
                tracing::warn!("Date type is not handled yet: {}", field.data_type());
                return None;
            }
        };

        Some(Self {
            time_column,
            time_format,
        })
    }

    #[allow(clippy::cast_possible_wrap)]
    #[allow(clippy::cast_possible_truncation)]
    pub(crate) fn convert(&self, timestamp_in_nanos: u128, op: Operator) -> Expr {
        let time_column: &str = self.time_column.as_ref();
        match &self.time_format {
            ExprTimeFormat::ISO8601 => binary_expr(
                cast(
                    col(time_column),
                    DataType::Timestamp(arrow::datatypes::TimeUnit::Millisecond, None),
                ),
                op,
                Expr::Literal(ScalarValue::TimestampMillisecond(
                    Some((timestamp_in_nanos / 1_000_000) as i64),
                    None,
                )),
            ),
            ExprTimeFormat::UnixTimestamp(format) => binary_expr(
                col(time_column),
                op,
                lit((timestamp_in_nanos / format.scale) as u64),
            ),
            ExprTimeFormat::Timestamp => binary_expr(
                col(time_column),
                op,
                Expr::Literal(ScalarValue::TimestampMillisecond(
                    Some((timestamp_in_nanos / 1_000_000) as i64),
                    None,
                )),
            ),
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use arrow::datatypes::{DataType, Field, TimeUnit};

    #[test]
    fn test_timestamp_filter_convert() {
        test(
            Field::new("timestamp", DataType::Int64, false),
            TimeFormat::UnixMillis,
            1_620_000_000_000_000_000,
            "timestamp > UInt64(1620000000000)",
        );
        test(
            Field::new("timestamp", DataType::Int64, false),
            TimeFormat::UnixSeconds,
            1_620_000_000_000_000_000,
            "timestamp > UInt64(1620000000)",
        );
        test(
            Field::new(
                "timestamp",
                DataType::Timestamp(TimeUnit::Second, None),
                false,
            ),
            TimeFormat::UnixSeconds,
            1_620_000_000_000_000_000,
            "timestamp > TimestampMillisecond(1620000000000, None)",
        );
        test(
            Field::new(
                "timestamp",
                DataType::Utf8,
                false,
            ),
            TimeFormat::UnixSeconds,
            1_620_000_000_000_000_000,
            "CAST(timestamp AS Timestamp(Millisecond, None)) > TimestampMillisecond(1620000000000, None)",
        );
    }

    fn test(field: Field, time_format: TimeFormat, timestamp: u128, expected: &str) {
        let time_column = "timestamp".to_string();
        let timestamp_filter_convert =
            TimestampFilterConvert::create(Some(field), Some(time_column), Some(time_format));
        assert!(timestamp_filter_convert.is_some());
        let timestamp_filter_convert =
            timestamp_filter_convert.expect("the convert can be created");
        let expr = timestamp_filter_convert.convert(timestamp, Operator::Gt);
        assert_eq!(expr.to_string(), expected);
    }
}
