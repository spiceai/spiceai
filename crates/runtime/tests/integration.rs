/*
Copyright 2024 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use std::{sync::Arc, time::Duration};

use arrow::array::RecordBatch;
use datafusion::{
    assert_batches_eq, execution::context::SessionContext,
    parquet::arrow::arrow_reader::ParquetRecordBatchReaderBuilder,
};
use futures::Future;
use runtime::{datafusion::DataFusion, Runtime};
use tracing::subscriber::DefaultGuard;
use tracing_subscriber::EnvFilter;

mod catalog;
mod docker;
mod federation;
mod graphql;
#[cfg(feature = "mysql")]
mod mysql;
#[cfg(feature = "postgres")]
mod postgres;
mod refresh_retry;
mod refresh_sql;
mod results_cache;
mod tls;

#[cfg(feature = "odbc")]
mod odbc;

/// Gets a test `DataFusion` to make test results reproducible across all machines.
///
/// 1) Sets the number of `target_partitions` to 3, by default its the number of CPU cores available.
fn get_test_datafusion() -> Arc<DataFusion> {
    let mut df = DataFusion::new();

    // Set the target partitions to 3 to make RepartitionExec show consistent partitioning across machines with different CPU counts.
    let mut new_state = df.ctx.state();
    new_state
        .config_mut()
        .options_mut()
        .execution
        .target_partitions = 3;
    let new_ctx = SessionContext::new_with_state(new_state);

    // Replace the old context with the modified one
    df.ctx = new_ctx.into();
    Arc::new(df)
}

fn init_tracing(default_level: Option<&str>) -> DefaultGuard {
    let filter = match (default_level, std::env::var("SPICED_LOG").ok()) {
        (_, Some(log)) => EnvFilter::new(log),
        (Some(level), None) => EnvFilter::new(level),
        _ => EnvFilter::new(
            "runtime=TRACE,datafusion-federation=TRACE,datafusion-federation-sql=TRACE",
        ),
    };

    let subscriber = tracing_subscriber::FmtSubscriber::builder()
        .with_env_filter(filter)
        .with_ansi(true)
        .finish();
    tracing::subscriber::set_default(subscriber)
}

async fn get_tpch_lineitem() -> Result<Vec<RecordBatch>, anyhow::Error> {
    let lineitem_parquet_bytes =
        reqwest::get("https://public-data.spiceai.org/tpch_lineitem.parquet")
            .await?
            .bytes()
            .await?;

    let parquet_reader =
        ParquetRecordBatchReaderBuilder::try_new(lineitem_parquet_bytes)?.build()?;

    Ok(parquet_reader.collect::<Result<Vec<_>, arrow::error::ArrowError>>()?)
}

type ValidateFn = dyn FnOnce(Vec<RecordBatch>);

async fn run_query_and_check_results<F>(
    rt: &mut Runtime,
    query: &str,
    expected_plan: &[&str],
    validate_result: Option<F>,
) -> Result<(), String>
where
    F: FnOnce(Vec<RecordBatch>),
{
    // Check the plan
    let plan_results = rt
        .datafusion()
        .ctx
        .sql(&format!("EXPLAIN {query}"))
        .await
        .map_err(|e| format!("query `{query}` to plan: {e}"))?
        .collect()
        .await
        .map_err(|e| format!("query `{query}` to results: {e}"))?;

    assert_batches_eq!(expected_plan, &plan_results);

    // Check the result
    if let Some(validate_result) = validate_result {
        let result_batches = rt
            .datafusion()
            .ctx
            .sql(query)
            .await
            .map_err(|e| format!("query `{query}` to plan: {e}"))?
            .collect()
            .await
            .map_err(|e| format!("query `{query}` to results: {e}"))?;

        validate_result(result_batches);
    }

    Ok(())
}

async fn wait_until_true<F, Fut>(max_wait: Duration, mut f: F) -> bool
where
    F: FnMut() -> Fut,
    Fut: Future<Output = bool>,
{
    let start = std::time::Instant::now();

    while start.elapsed() < max_wait {
        if f().await {
            return true;
        }

        tokio::time::sleep(Duration::from_millis(100)).await;
    }

    false
}

fn container_registry() -> String {
    std::env::var("CONTAINER_REGISTRY")
        .unwrap_or_else(|_| "public.ecr.aws/docker/library/".to_string())
}
