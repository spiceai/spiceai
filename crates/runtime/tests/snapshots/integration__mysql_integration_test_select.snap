---
source: crates/runtime/tests/integration.rs
description: "Query: SELECT * FROM test"
---
+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |
+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | BytesProcessedNode                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
|               |   Federated                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
|               |  Projection: test.id, test.col_bit, test.col_tiny, test.col_short, test.col_long, test.col_longlong, test.col_float, test.col_double, test.col_timestamp, test.col_date, test.col_time, test.col_blob, test.col_varchar, test.col_string, test.col_var_string, test.col_decimal, test.col_unsigned_int, test.col_char, test.col_set, test.col_json                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
|               |   TableScan: test                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| physical_plan | BytesProcessedExec                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
|               |   SchemaCastScanExec                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |
|               |     VirtualExecutionPlan name=mysql compute_context=host=localhost,port=13316,db=mysqldb,user=root sql=SELECT test.id, test.col_bit, test.col_tiny, test.col_short, test.col_long, test.col_longlong, test.col_float, test.col_double, test.col_timestamp, test.col_date, test.col_time, test.col_blob, test.col_varchar, test.col_string, test.col_var_string, test.col_decimal, test.col_unsigned_int, test.col_char, test.col_set, test.col_json FROM test rewritten_sql=SELECT `test`.`id`, `test`.`col_bit`, `test`.`col_tiny`, `test`.`col_short`, `test`.`col_long`, `test`.`col_longlong`, `test`.`col_float`, `test`.`col_double`, `test`.`col_timestamp`, `test`.`col_date`, `test`.`col_time`, `test`.`col_blob`, `test`.`col_varchar`, `test`.`col_string`, `test`.`col_var_string`, `test`.`col_decimal`, `test`.`col_unsigned_int`, `test`.`col_char`, `test`.`col_set`, `test`.`col_json` FROM `test` |
|               |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
